Add disk usage summary report interface

dotnwat opened the pull request:
In a previous commit the `gc_estimate` interface was introduced which was effectively a dry-run of retention policy designed to allow future disk space management functionality to reason about which partitions had the most data available for removal from local disk in support of low disk space scenarios.

The flip side of knowing how much data can be reclaimed is to know how much data is being used. For example, future disk space management functionality will need to understand how much space is attributable to a raft vs other sub-systems (e.g. in single-disk setups).

To that end this PR introduces `disk_log_impl::disk_usage()` that replaces `gc_estimate` and returns a report about both disk usage broken out into categories (e.g. data, index) as well as the previous dry-run retention calculation. They are combined because (1) it is more efficient to collect these statistics at once and (2) the only current user will need both pieces of information.

```
ss::future<usage_report> disk_log_impl::disk_usage(compaction_config cfg) 

/*
 * disk usage report
 *
 * usage: disk usage summary for log.
 * reclaim: disk usage reclaim summary for log.
 */
struct usage_report {
    usage usage;
    reclaim_size_limits reclaim;
};
```

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.1.x
- [ ] v22.3.x
- [ ] v22.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
* none

dotnwat commented:
Failures:
* https://github.com/redpanda-data/redpanda/issues/7756

VladLazar commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1750,28 +1750,50 @@ disk_log_impl::estimate_reclaim_size(compaction_config cfg) {
+        usage{},
+        [](usage acc, usage u) { return acc + u; }),
+
+      // reduce segments available for reclaim
```
This comment should say something like: `reduce segments not available for reclaim`.

VladLazar reviewed (approved):
Looks good. Only nits and suggestions.

dotnwat commented on src/v/storage/segment.cc:
```diff
@@ -162,9 +174,17 @@ segment::remove_persistent_state(std::filesystem::path path) {
     co_return 0;
 }
 
+void segment::clear_cached_disk_usage() {
+    _idx.clear_cached_disk_usage();
+    _data_disk_usage_size.reset();
+    _compaction_index_size.reset();
+}
```
right. i tried very hard (and succeeded) in not refactoring a bunch of stuff to make this cleaner. it is indeed awkward to have the reader and appender and then the same path in the segment/reader/appender etc...

andrwng commented on src/v/storage/segment.h:
```diff
@@ -166,7 +166,7 @@ class segment {
      * removed already (i.e. marked as a tombstone and closed) then the total
```
It looks like this bit about behavior after segment removal may no longer be true?

dotnwat replied:
fixed

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1701,15 +1701,10 @@ log make_disk_backed_log(
 
 ss::future<reclaim_size_limits>
 disk_log_impl::estimate_reclaim_size(compaction_config cfg) {
```
> However, even when retention is disabled we may still want to delete data that has been uploaded to cloud storage.

Just making sure I understand the behavior before this: did we previously not locally GC for compacted topics? Or did we just not estimate the size, and therefore not prioritize running GC code as highly as we should've been

dotnwat replied:
Before there was no estimation. Now we can estimate, but we aren't yet using that in the scheduling decision. Also, generally, this is work is tackling _retention_ policies, so it is orthogonal to compaction. We aren't estimating for compaction.

andrwng replied:
Got it, thanks for clarifying

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1733,15 +1728,24 @@ disk_log_impl::estimate_reclaim_size(compaction_config cfg) {
     fragmented_vector<segment_set::type> available_segments;
     for (auto& seg : _segs) {
-        if (seg->offsets().dirty_offset <= retention_offset) {
+        if (
```
done!

dotnwat commented:
Force pushes
* https://github.com/redpanda-data/redpanda/compare/5f461ac13ebacac502ab90d876bd2f042455b38d..fc547bc2ff5663fbd2222a81a7afa06e0025723e
  * Updated comment
  * Added test for non-collectible log
* https://github.com/redpanda-data/redpanda/compare/fc547bc2ff5663fbd2222a81a7afa06e0025723e..41ddc8d70e0efdfd429f535b1add8a579ba8d43a
  * Updated comment
